text/2020-03-12-load-based-region-splitting.md

```diff
@@ -0,0 +1,80 @@
+# RFC: Load-based Region Splitting
+
+## Summary
+
+This proposal presents a load-based region splitting method that automatically
+splits region when it remains a read hotspot continuously. After splitting, the
+hotspot can be easily scattered through Region scheduling.
+
+## Motivation
+
+The existing scheduling mechanism (`hot-region-scheduler` in TiKV 3.0) supports
+distributing multiple hot read regions on the same node to reduce the load on
+the node. However, if all read hotspots are in the same region, this mechanism
+cannot further distribute hotspots. A temporary approach is to manually split
+the region and suspend the region merge, but its operation cost is high and it
+has negative impacts on non-hotspots.
+
+CockroachDB introduced Load-Based Splitting to solve this problem. After
+investigation, we think this method is also applicable to TiKV clusters, so
+reference will be made to its functional design.
+
+## Design
+
+### Design Considerations
+
+* There is a cost to perform a split. And after splitting, hotspots can only be
+  solved after the leaders of newly generated multiple regions have been
+  distributed to different nodes. In this case, splitting is meaningless for
+  momentary and short  loads（< 10s). Therefore, load-based splitting should be
+  performed only after intense traffic generating in the region lasts for a
+  period of time.
+* After Region has been split into two new regions, the split regions should be
+  prevented from being merged immediately. Therefore, the overall loads of the
+  two regions should be checked before the region merge.
+* To avoid back-and-forth splitting and merging, the merging load threshold
+  should be slightly lower than splitting load threshold (e.g., 20% lower).
+
+### Detailed Design
+
+1. When the read load of a region exceeds the splitting threshold, start
+   counting the traffic of keys. Through probability sampling of requests on
+   this region, 20 keys are randomly selected first.
```

Comment:
`20 keys are randomly selected` i don't know if there is a efficient way to do this.

Reply:
@zhangjinpeng1987 

if there are N keys accessed in the first second, wen can find the 20 keys with following method, for the most case, it's complexity is O(1), for the worst case, it's O(N)

```python
keys = [...]
result = set([])
for i in range(20):
    idx = random(0, N)
    idx2 = idx
    while True:
        if keys[idx2] not in result:
            result.add(keys[idx2])
            break
        idx2 = (idx2 + 1) % N
        if idx2 == idx:
            return result # not enough keys
return result       
```

if the worst case is not acceptable, we can limit the scan size to 100(or even 1), but  we'll have a larger chance for not having enough keys. And I think less than 20 keys is not a problem.

```python
keys = [...]
result = set([])
for i in range(20):
    idx = random(0, N)
    idx2 = idx
    for j in range(100):
        idx2 = (idx + j) % N
        if keys[idx2] not in result:
            result.add(keys[idx2])
            break
return result       
```

Reply:
We sampled the keys by 4MB distance and store them in the SST's user properties, we can say these keys split the region into many small pieces, I think we can make use of these keys to speed up the selection process.